### PR TITLE
[DOP-4438]: Add component filtering for production builds in the Autobuilder

### PR DESCRIPTION
**Pull request by branberry**
### Stories/Links:

[DOP-4438](https://jira.mongodb.org/browse/DOP-4438)

### Notes

This PR adds the component factory filter functionality to builds.

[successful build](https://workerpoolstaging-qgeyp.mongodbstitch.com/pages/job.html?collName=queue&jobId=660edc0b3a5920a2911819a1)

### README updates

- - [ ] This PR introduces changes that should be reflected in the README, and I have made those updates.
- - [x] This PR does not introduce changes that should be reflected in the README

### CODE REVIEW

File: tests/unit/job/productionJobHandler.test.ts [hunk 1]
```diff
@@ -244,12 +238,6 @@ describe('ProductionJobHandler Tests', () => {
     jobHandlerTestHelper.setupForSuccess();
     await jobHandlerTestHelper.jobHandler.execute();
     jobHandlerTestHelper.verifyNextGenSuccess();
-
-    expect(jobHandlerTestHelper.fileSystemServices.writeToFile).toBeCalledWith(
-      `repos/${jobHandlerTestHelper.job.payload.repoName}/.env.production`,
```
Comment (branberry) [inline tests/unit/job/productionJobHandler.test.ts:249]: Removed this as I believe it's not providing a ton of value, and leads to additional overhead when making changes.

File: tests/unit/job/productionJobHandler.test.ts [hunk 2]
```diff
@@ -166,12 +166,6 @@ describe('ProductionJobHandler Tests', () => {
     jobHandlerTestHelper.setupForSuccess();
     await jobHandlerTestHelper.jobHandler.execute();
     jobHandlerTestHelper.verifyNextGenSuccess();
-    // TODO: Correct number of arguments
-    expect(jobHandlerTestHelper.fileSystemServices.writeToFile).toBeCalledWith(
-      `repos/${jobHandlerTestHelper.job.payload.repoName}/.env.production`,
```
Comment (branberry) [inline tests/unit/job/productionJobHandler.test.ts:171]: Removed this as I believe it's not providing a ton of value, and leads to additional overhead when making changes.

File: Dockerfile [hunk 3]
```diff
@@ -63,17 +63,30 @@ USER docsworker-xlarge
...
+ENV PATH="${WORK_DIRECTORY}/.cargo/bin:${PATH}"
+RUN echo ${WORK_DIRECTORY}
+
 # get shared.mk
 RUN curl https://raw.githubusercontent.com/mongodb/docs-worker-pool/meta/makefiles/shared.mk -o shared.mk
 
 # install snooty frontend and docs-tools
 RUN git clone -b v${SNOOTY_FRONTEND_VERSION} --depth 1 https://github.com/mongodb/snooty.git       \
     && cd snooty                                                                                   \
-    && npm ci --legacy-peer-deps --omit=dev                                                        \
```
Comment (branberry) [inline Dockerfile:72]: Need to remove omit dev as the filter functionality depends on a dev dependency.

Comment (mmeigs) [reply Dockerfile:72]: Wondering why it's a dev dependency then?

File: src/job/jobHandler.ts [hunk 4]
```diff
@@ -401,6 +401,12 @@ export abstract class JobHandler {
       envVars += `PATH_PREFIX=${pathPrefix}\n`;
     }
     const snootyFrontEndVars = {
+      // Setting this to always be true. It's one less SSM parameter to manage,
+      // and it doesn't hurt to always do the filtering. The only reason
+      // it exists here is so that we don't always filter when testing locally.
+      // This is because the filter functionality updates the ComponentFactory.js file
+      // which could be accidentally committed if run locally.
+      USE_FILTER_BRANCH: true,
```
Comment (mmeigs) [inline src/job/jobHandler.ts:409]: Will there be an easy way to flip this variable if something goes wrong in prod though? Just wondering on a quick roll-back plan.

**Review by mmeigs — APPROVED** [review]

**Review by anabellabuckvar — APPROVED** [review]
LGTM!

**Review by rayangler — APPROVED** [review]
LGTM! Reviewed live with @branberry

File: Dockerfile (hunk 3 above)
Comment (rayangler) [reply Dockerfile:72]: I had the same question! We should definitely consider moving the relevant dependencies to be installed for prod in the near future if possible

**Pull request merged**